include/render.h

```diff
@@ -94,17 +96,75 @@ struct Render_t {
 	bool fullFrame = true;
 };
 
+struct RenderedImage {
+	uint16_t width              = 0;
+	uint16_t height             = 0;
+	bool double_width           = false;
+	bool double_height          = false;
+	bool flip_vertical          = false;
+	Fraction pixel_aspect_ratio = {};
+	uint8_t bits_per_pixel      = 0;
+	uint16_t pitch              = 0;
+	const uint8_t* image_data   = nullptr;
+	const uint8_t* palette_data = nullptr;
+
+	inline bool is_paletted() const
+	{
+		return (bits_per_pixel == 8);
+	}
+
+	RenderedImage deep_copy() const
+	{
+		RenderedImage copy = *this;
+
+		// Deep-copy image and palette data
+		const auto image_data_num_bytes = height * pitch;
+
+		copy.image_data = static_cast<uint8_t*>(
+		        std::malloc(image_data_num_bytes));
+		assert(copy.image_data);
+
+		assert(image_data);
+		std::memcpy(const_cast<uint8_t*>(copy.image_data),
+		            image_data,
+		            image_data_num_bytes);
+
+		// TODO it's bad that we need to make this assumption downstream
+		// on the size and alignment of the palette...
+		if (palette_data) {
+			constexpr auto PaletteNumBytes = 256 * 4;
+
+			copy.palette_data = static_cast<uint8_t*>(
+			        std::malloc(PaletteNumBytes));
+			assert(copy.palette_data);
+
```

Comment:
Suggest `new` because it throws on an allocation exception (no need to catch it; it'll take down the program if the host really is out of memory), so you're guaranteed to have it and can skip the assert.

Reply:
Sure. Although then all bets are off if we're out of mem 😏

Reply:
Done.